### PR TITLE
feat: Calendar Cache

### PR DESCRIPTION
**Pull request by zomars**
## What does this PR do?

- Implements a basic third-party calendar cache that lasts for a minute (for v1)
- First call will be as slow as now, but subsequent calls should be nearly instant.
- Migrates gCal CalendarService from callbacks to async/awaits

## Possible downsides

- If a profile is a high traffic one, we might run into a race condition where the cache is de-synced with the actual availability and get a "Could not book the meeting." error.

## TODO

- [x] Implement a cron for cleanup expired caches
- [x] Implement an opt-in feature mechanism 
- [ ] Implement a way to bypass cache altogether (either via request header or search parameter)
- [x] Make the cache more aggressive (save full months, even if only a day was requested)

## Type of change

- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## How should this be tested?

- Run migrations
- Install Google Calendar
- Enable some calendars
- As an ADMIN, enable the calendar-cache feature flag
- Go to a booking page
- Check if CalendarCache table is being populated correctly
- Check availability works as normal

## Mandatory Tasks

- [x] Make sure you have self-reviewed the code. A decent size PR without self-review might be rejected.

### CODE REVIEW

**Comment by PeerRich** [comment]
ouhh exciting

**Comment by emrysal** [comment]
The problem is that the current bottleneck is not the calendars - the bottleneck is the database right now 😓

**Comment by zomars** [comment]
> The problem is that the current bottleneck is not the calendars - the bottleneck is the database right now 😓

This is not an attempt to fix that bottleneck. This is another mechanism altogether. But agreed.

**Review by zomars — COMMENTED** [review]
Ready for review and manual testing

**Comment by PeerRich** [comment]
> If a profile is a high traffic one, we might run into a race condition where the cache is de-synced with the actual availability and get a "Could not book the meeting." error.


for this, cant we use our reservation queue system?

we need to prevent "Could not book the meeting." errors at all cost

**Comment by zomars** [comment]
>  > If a profile is a high traffic one, we might run into a race condition where the cache is de-synced with the actual availability and get a "Could not book the meeting." error.
> 
> 
> for this, cant we use our reservation queue system?
> 
> we need to prevent "Could not book the meeting." errors at all cost

Could alleviate but we can't really guarantee

**Review by zomars — COMMENTED** [review]
I do think this is ready to merge. Even if we deploy to prod it won't be active until we activate it via feature flag.

File: apps/web/vercel.json [hunk 1]
```diff
@@ -0,0 +1,8 @@
+{
+  "crons": [
+    {
+      "path": "/api/cron/calendar-cache-cleanup",
+      "schedule": "0 5 * * *"
```
Comment (alannnc) [inline apps/web/vercel.json:5]:
```suggestion
      // Every day At 05:00.” 
      "schedule": "0 5 * * *"
```

Comment (alannnc) [reply apps/web/vercel.json:5]: A little hint to the eye

File: packages/app-store/googlecalendar/lib/CalendarService.ts [hunk 2]
```diff
@@ -24,13 +25,45 @@ interface GoogleCalError extends Error {
   code?: number;
 }
 
+const ONE_MINUTE_MS = 60 * 1000;
+const CACHING_TIME = ONE_MINUTE_MS;
+
+/** Expand the start date to the start of the month */
+function getTimeMin(timeMin: string) {
```
Comment (alannnc) [inline packages/app-store/googlecalendar/lib/CalendarService.ts:32]:
```suggestion
function getStartDateOfTheMonth(timeMin: string) {
```

File: packages/app-store/googlecalendar/lib/CalendarService.ts [hunk 3]
```diff
@@ -24,13 +25,45 @@ interface GoogleCalError extends Error {
...
+  const dateMin = new Date(timeMin);
+  return new Date(dateMin.getFullYear(), dateMin.getMonth(), 1, 0, 0, 0, 0).toISOString();
+}
```
Comment (alannnc) [inline packages/app-store/googlecalendar/lib/CalendarService.ts:35]:
```suggestion
function getStartDateOfTheMonth(dateStart: string) {
  const dateMin = new Date(dateStart);
  return new Date(dateMin.getFullYear(), dateMin.getMonth(), 1, 0, 0, 0, 0).toISOString();
}
```

File: packages/app-store/googlecalendar/lib/CalendarService.ts [hunk 4]
```diff
@@ -24,13 +25,45 @@ interface GoogleCalError extends Error {
...
+
+/** Expand the end date to the end of the month */
+function getTimeMax(timeMax: string) {
+  const dateMax = new Date(timeMax);
+  return new Date(dateMax.getFullYear(), dateMax.getMonth() + 1, 0, 0, 0, 0, 0).toISOString();
+}
```
Comment (alannnc) [inline packages/app-store/googlecalendar/lib/CalendarService.ts:41]:
```suggestion
function getEndTimeOfTheMonth(endTime: string) {
  const dateMax = new Date(endTime);
  return new Date(dateMax.getFullYear(), dateMax.getMonth() + 1, 0, 0, 0, 0, 0).toISOString();
}
```

File: packages/app-store/googlecalendar/lib/CalendarService.ts [hunk 5]
```diff
@@ -24,13 +25,45 @@ interface GoogleCalError extends Error {
...
+ * Enable or disable the expanded cache
+ * TODO: Make this configurable
+ * */
+const ENABLE_EXPANDED_CACHE = true;
+
+/**
+ * By expanding the cache to whole months, we can save round trips to the third party APIs.
+ * In this case we already have the data in the database, so we can just return it.
+ */
+function handleMinMax(min: string, max: string) {
```
Comment (alannnc) [inline packages/app-store/googlecalendar/lib/CalendarService.ts:53]:
```suggestion
function handleStartAndEndOfMonth(min: string, max: string) {
```

File: packages/app-store/googlecalendar/lib/CalendarService.ts [hunk 6]
```diff
@@ -96,345 +138,340 @@ export default class GoogleCalendarService implements Calendar {
...
-        payload["conferenceData"] = calEventRaw.conferenceData;
-      }
-      const calendar = google.calendar({
-        version: "v3",
+    if (calEventRaw.conferenceData && calEventRaw.location === MeetLocationType) {
+      payload["conferenceData"] = calEventRaw.conferenceData;
+    }
+    const calendar = await this.authedCalendar();
+    // Find in calEventRaw.destinationCalendar the one with the same credentialId
+
```
Comment (alannnc) [inline packages/app-store/googlecalendar/lib/CalendarService.ts:183]:
```suggestion
    
    // Find in calEventRaw.destinationCalendar the one with the same credentialId
```

File: packages/app-store/googlecalendar/lib/CalendarService.ts [hunk 7]
```diff
@@ -96,345 +138,340 @@ export default class GoogleCalendarService implements Calendar {
...
-          eventId: uid,
-          sendNotifications: false,
-          sendUpdates: "none",
+    if (cached) return cached.value as unknown as calendar_v3.Schema$FreeBusyResponse;
+
+    const apires = await calendar.freebusy.query({
+      requestBody: { timeMin, timeMax, items },
+    });
+
+    await prisma.calendarCache.upsert({
```
Comment (alannnc) [inline packages/app-store/googlecalendar/lib/CalendarService.ts:389]: Can we omit this await? To make it respond faster.

**Review by alannnc — DISMISSED** [review]
Code looks good! Left some NIT changes. Still since it's moving important calendar service methods we still need to test it.

File: apps/web/vercel.json (hunk 1 above)
Comment (zomars) [reply apps/web/vercel.json:5]: I don't think we can put comments on naked json files

File: packages/app-store/googlecalendar/lib/CalendarService.ts [hunk 8]
```diff
@@ -129,306 +170,302 @@ export default class GoogleCalendarService implements Calendar {
...
+    try {
+      const event = await calendar.events.insert({
+        calendarId: selectedCalendar,
+        requestBody: payload,
+        conferenceDataVersion: 1,
+        sendUpdates: "none",
+      });
+
+      if (event && event.data.id && event.data.hangoutLink) {
+        await calendar.events.patch({
```
Comment (zomars) [inline packages/app-store/googlecalendar/lib/CalendarService.ts:214]: There was a missing await here.

File: packages/app-store/googlecalendar/lib/CalendarService.ts [hunk 9]
```diff
@@ -129,306 +170,302 @@ export default class GoogleCalendarService implements Calendar {
...
-          if (err) {
-            console.error("There was an error contacting google calendar service: ", err);
-            return reject(err);
-          }
-
-          if (evt && evt.data.id && evt.data.hangoutLink && event.location === MeetLocationType) {
-            calendar.events.patch({
-              // Update the same event but this time we know the hangout link
-              calendarId: selectedCalendar,
-              auth: myGoogleAuth,
```
Comment (zomars) [inline packages/app-store/googlecalendar/lib/CalendarService.ts:271]: calendar is already authed

File: apps/web/vercel.json (hunk 1 above)
Comment (zomars) [reply apps/web/vercel.json:5]: image

File: packages/app-store/googlecalendar/lib/CalendarService.ts (hunk 3 above)
Comment (zomars) [reply packages/app-store/googlecalendar/lib/CalendarService.ts:35]: Will keep timeMin/Max to match what gCal is expecting.

**Review by zomars — COMMENTED** [review]
Self review done

File: packages/app-store/googlecalendar/lib/CalendarService.test.ts [hunk 10]
```diff
@@ -0,0 +1,99 @@
+import { afterEach, expect, test, vi } from "vitest";
+
+import prismaMock from "../../../../tests/libs/__mocks__/prisma";
```
Comment (emrysal) [inline packages/app-store/googlecalendar/lib/CalendarService.test.ts:3]: lil lint fix needed

**Review by emrysal — APPROVED** [review]
It looks good to me :)

**Pull request merged**